radicle-httpd/src/api/v1/projects.rs

```diff
@@ -0,0 +1,331 @@
+                } else if let Some(since) = since {
+                    q.committer.time.seconds() >= since
+                } else if let Some(until) = until {
+                    q.committer.time.seconds() < until
+                } else {
+                    // If neither `since` nor `until` are specified, we include the commit.
+                    true
+                }
+            } else {
+                false
+            }
+        })
+        .skip(page * per_page)
+        .take(per_page)
+        .filter_map(|commit| {
+            if let Ok(commit) = commit {
+                radicle_surf::commit(&repo.raw().into(), commit.id).ok()
+            } else {
+                None
+            }
+        })
+        .collect::<Vec<_>>();
+
+    let response = json!({
+        "headers": headers,
+        "stats":  stats(&repo)?,
+    });
+
+    if fallback_to_head {
+        return Ok::<_, Error>((StatusCode::FOUND, Json(response)));
+    }
+
+    Ok::<_, Error>((StatusCode::OK, Json(response)))
+}
+
+/// Get project commit.
+/// `GET /projects/:project/commits/:sha`
+async fn commit_handler(
+    Extension(ctx): Extension<Context>,
+    Path((project, sha)): Path<(Id, Oid)>,
+) -> impl IntoResponse {
+    let storage = &ctx.profile.storage;
+    let repo = storage.repository(project)?;
+    let commit = radicle_surf::commit(&repo.raw().into(), sha)?;
+
+    Ok::<_, Error>(Json(commit))
+}
+
+/// Get project activity for the past year.
+/// `GET /projects/:project/activity`
+async fn activity_handler(
+    Extension(ctx): Extension<Context>,
+    Path(project): Path<Id>,
+) -> impl IntoResponse {
+    let current_date = chrono::Utc::now().timestamp();
+    let one_year_ago = chrono::Duration::weeks(52);
+    let storage = &ctx.profile.storage;
+    let repo = storage.repository(project)?;
+    let (_, head) = repo.head()?;
+    let timestamps = History::new(repo.raw().into(), head)?
+        .filter_map(|a| {
+            if let Ok(a) = a {
+                let seconds = a.committer.time.seconds();
+                if seconds > current_date - one_year_ago.num_seconds() {
+                    return Some(seconds);
+                }
+
+                None
+            } else {
+                None
+            }
+        })
+        .collect::<Vec<i64>>();
+
+    Ok::<_, Error>((StatusCode::OK, Json(json!({ "activity": timestamps }))))
+}
+
+/// Get project source tree.
+/// `GET /projects/:project/tree/:sha/*path`
+async fn tree_handler(
+    Extension(ctx): Extension<Context>,
+    Path((project, sha, path)): Path<(Id, Oid, String)>,
+) -> impl IntoResponse {
+    let path = path.strip_prefix('/').ok_or(Error::NotFound)?.to_string();
+    let storage = &ctx.profile.storage;
+    let repo = storage.repository(project)?;
+    let tree = radicle_surf::object::tree(&repo.raw().into(), Some(Sha { sha }), Some(path))?;
+    let response = json!({
+        "path": &tree.path,
+        "entries": &tree.entries,
```

Comment:
would be nice if `radicle_surf::Revision` implements `From<Oid>`

Reply:
I'm actually going to remove `radicle_surf::Revision` and use the `radicle_surf::git::Revision` trait instead. This means you can use the plain `Oid`, or anything that implements that trait, which a lot of types do :)